### PR TITLE
lib.path.subpath.join: init

### PR DESCRIPTION
**Pull request by infinisil**
###### Description of changes

This function can be used to safely join subpaths together, a bit like `concatStringsSep "/"`, but safer. Relates to other work in [the path library effort](https://github.com/NixOS/nixpkgs/issues/210426).

This work is sponsored by [Antithesis](https://antithesis.com/) :sparkles:

###### Things done

- [x] Wrote documentation
- [x] Added and ran tests

### CODE REVIEW

File: lib/path/tests/prop.nix
Comment (fricklerhandwerk) [inline lib/path/tests/prop.nix:99]: This is a bit ugly. Given that we will extend the test suite anyway (I haven't looked at the other PRs yet), maybe we should do a little bit of generalisation. I feel uneasy about the change to the randomly generated paths, because it makes the whole setup more complicated and significantly harder to understand before getting into the actual testing. Have you considered approaching this differently? Instead of producing lists of paths in the random generator, produce a sequence of random numbers (which can be reused for other purposes), pass that and do the chunking in Nix.

Maybe we could also have one Nix file per test, and import each from an entry point that does the pre-processing of the input? Then each file can take whatever data type it cares about, directly. The main file will then only combine the tests in a way that produces output as needed for further processing back in the shell. I suppose, with the other tests, we will have a series of asserts and then produce a value, but having that in one place without the other noise will allow for easier changes of factoring.

File: lib/path/tests/generate.awk
Comment (fricklerhandwerk) [inline lib/path/tests/generate.awk:34]: I'm really not fond of this extra complexity, see the comment at the end of the Nix file.

File: lib/path/default.nix
Comment (fricklerhandwerk) [inline lib/path/default.nix:210]:
```suggestion
          then throw ''
            lib.path.subpath.join:
              Element at index ${toString i} is not a valid subpath string:
                ${subpathInvalidReason path}
          ''
```

Since those strings are fairly long, I think we should break them down to roughly fit into a standard terminal.
No strong opinion on indentation, but it *may* help with debugging by indicating level of detail.

File: lib/path/default.nix
Comment (roberth) [reply lib/path/default.nix:203]: We shouldn't use redundant parentheses here, because if anyone doubts their understanding of the language, they can use common sense to infer how the operator precedence works here.
Yet, if someone sees redundant parentheses, they may start to think that their understanding of the language isn't good, even if they're already quite productive in it.

File: lib/path/tests/prop.nix
Comment (roberth) [reply lib/path/tests/prop.nix:99]: It would be nice not to have the generation logic in Nix expressions, to decouple the Nix tests from the test runner. However, we also use this pattern for invoking external commands that we model in Nix, such as `realpath`, so we can't easily get away from all coupling between framework and tests.

I do think it's worth exploring for the sake of writing Nix-only tests in a consistent style where all domain-specific testing logic is in the same language.
Instead of generating examples or a large chunk of randomness ahead of time, we could easily create a pseudorandom number generator by invoking `builtins.hashString "sha256"` repeatedly on a state variable, starting from a small seed that's generated outside of Nix. This achieves the goal without ever running out of randomness, and therefore without any coupling with the non-Nix world. The state variable can be handled by a simplified Nix equivalent of the [quickcheck `Gen`](https://hackage.haskell.org/package/QuickCheck-2.14.2/docs/Test-QuickCheck-Gen.html) library. Instead of applicative we could have a single function that takes an attrset of "gens" and a function matching those attr names, etc.

Comment (fricklerhandwerk) [reply lib/path/tests/prop.nix:99]: I don't agree we should push toward implementing more things in the Nix language, even if it's possible and not all that hard to do. It will happen anyway, but should happen much more slowly and deliberately. That is, I wouldn't oppose a QuickCheck implementation or PRs to migrate existing tests to it. But for the sake of testing this library, simple Nix expressions that batch-process inputs should do.

File: lib/path/default.nix
Comment (infinisil) [reply lib/path/default.nix:150]: Not really possible since it's not rendered as markdown and we don't know the link of this manual in the end. I'd like to defer this to when we have proper markdown rendering

File: lib/path/default.nix
Comment (infinisil) [reply lib/path/default.nix:166]: https://github.com/NixOS/nixpkgs/pull/209099#discussion_r1072576833

File: lib/path/default.nix
Comment (infinisil) [reply lib/path/default.nix:203]: I like the `subpathInvalidReason` call a bit more over `isValid` because:
- `subpathInvalidReason` is used again later in this function, so we can't remove the dependency on it anyways, and it's nice to see the correspondence of the two instances.
- Currently the public functions of this library only depend on the internal functions. This is because generally the public functions have some extra error checking which isn't present in the internal ones. While `isValid` doesn't have any such error checking, it's still nice to maintain this property imo

File: lib/path/default.nix
Comment (infinisil) [reply lib/path/default.nix:208]: Some personal feedback: I think minor naming changes of internal variables is bikeshedding, and it's probably better to not give such feedback to both save me and you time :sweat_smile:

File: lib/path/default.nix
Comment (infinisil) [reply lib/path/default.nix:216]: I don't agree with making this error even more verbose. I don't expect this code to be updated more than once a decade, and once it does we have extensive checks which should catch such bugs fairly easily. In the exceptional case where a contributor encounters this error they can still just file an issue and the ping will reach us somehow in any case (git log, PR searches, etc.).

File: lib/path/default.nix
Comment (infinisil) [reply lib/path/default.nix:209]: https://github.com/NixOS/nixpkgs/pull/209099#discussion_r1072589152

File: lib/path/default.nix
Comment (infinisil) [reply lib/path/default.nix:218]: I like the comment here, `seq` is not well known and we should justify its use

File: lib/path/tests/prop.sh
Comment (infinisil) [reply lib/path/tests/prop.sh:65]: Good idea, doing that

File: lib/path/default.nix
Comment (infinisil) [reply lib/path/default.nix:210]: See https://github.com/NixOS/nixpkgs/pull/208887#discussion_r1069737602

File: lib/path/tests/prop.nix
Comment (infinisil) [reply lib/path/tests/prop.nix:99]: Since this is test code, we could maybe even allow using the cursed `builtins.exec`:

```
$ nix-instantiate --eval -E 'builtins.exec [ "realpath" "-m" "//foo/./bar/." ]' --allow-unsafe-native-code-during-evaluation
/foo/bar
```

But no I don't think it would be a great idea to write all tests in Nix itself. Nix is just not made for this kind of thing. One of the main restrictions is that we can't check for all kinds of evaluation failures. `builtins.tryEval` only works for `throw` and `assert`, no other failures, and it doesn't allow checking for the error message either. Randomness too is not something Nix was made for. Yes we can hack around it with `builtins.hashString`, custom PRNG's (I even wrote one in Nix myself some time ago) and monadic generator passing, but that's gonna be painful.

I think something like https://github.com/Mic92/pythonix is the right way to go. Unfortunately that project is archived now, but it works by linking to `libexpr` directly and using that to evaluate expressions. This way failures can be caught and checked. The main problem with this is that `libexpr` isn't stable, so it requires continuous maintenance. But using that approach testing libraries of generic programming languages can be used.

The alternative of just using the CLI as a stable API has the problem that it's way too slow (this is why I opted for just a single CLI call in these tests here).

I'll see if I can do some more decoupling for the property tests here though. Maybe having a separate file/folder for every property to check and using the `.nix` file only for batch evaluation and no checking.

File: lib/path/default.nix
Comment (fricklerhandwerk) [reply lib/path/default.nix:203]: > it's nice to see the correspondence of the two instances.

> Currently the public functions of this library only depend on the internal functions. This is because generally the public functions have some extra error checking which isn't present in the internal ones. 

I find these rather weak arguments in themselves, but most importantly they miss my point.
Reusing the function obscures the purpose of the check, and reveals internals that are irrelevant at that point, and distracting. All we care about there is whether the path is valid, not how validity is determined (the error is null).

However internal a given piece of code may seem, it's still primarily for people to read, therefore these little things matter a lot. Every time someone has to complete your intended meaning in their head it's wasting (highly constrained) mental resources on something that is not "actually understanding what you're trying to achieve here".

I know that sounds really nitpicky, but I'm also convinced we've achieved the goal of highly maintainable code when someone as lazy and distracted as myself can read through the code linearly without much effort and still be sure to understand it completely. That may be a highly subjective and to often unsuitable measure, but it's the closest to thing I have to imagine how uninitiated readers will deal with the code.

Comment (infinisil) [reply lib/path/default.nix:203]: Yeah agree that this sounds better, I'm changing it to `isValid` throughout the code

File: lib/path/tests/prop.nix
Comment (infinisil) [reply lib/path/tests/prop.nix]: I think it would be better to just drop the property tests from this PR. I don't want to have this be blocked on improved property tests. There's still unit tests as well.

**Comment by infinisil** [comment]
@fricklerhandwerk @roberth I'm now working on https://github.com/NixOS/nix/pull/7735 to have better Nix testing support in the future. For now I removed the more hacky property tests from this PR so it's not [blocked on that](https://github.com/NixOS/nixpkgs/pull/209099#discussion_r1069177623), there's still unit tests in any case.

I also rebased this PR after the merge of #208887, to me this PR looks good like this

File: lib/path/default.nix
Comment (infinisil) [inline lib/path/default.nix:259]: I simplified this assertion a lot by introducing `lib.asserts.assertAll`, now included in this PR

Comment (infinisil) [reply lib/path/default.nix]: Never mind, I split that function into https://github.com/NixOS/nixpkgs/pull/215166

File: lib/path/default.nix
Comment (infinisil) [inline lib/path/default.nix:267]: I also simplified this a bit to remove the `seq` and `abort` fallthru thing, I think that was overengineered.

Comment (roberth) [reply lib/path/default.nix:267]: I really don't like the abuse of the exception "side effect" here.
It starts with `assert`, which is never triggered and then it continues with `assertMsg`, which is also never triggered. The final error happens to be in the error message of `assertMsg`, evaluated when it almost triggers.
This is enough reason to just use an `if` expression instead.

Final nitpick: assertions are for checking the invariants of self-contained software units, and not for input validation across a public interface. See [assert (verb)](https://en.wiktionary.org/wiki/assert)
> 1\. To [declare](https://en.wiktionary.org/wiki/declare) with [assurance](https://en.wiktionary.org/wiki/assurance) or plainly and strongly; to state positively.
> [...]
> (programming) To specify that a condition or expression is true at a certain point in the code.

Comment (infinisil) [reply lib/path/default.nix]: Good point, using an `if` now.

Regarding assertions in general, yeah maybe `assert` is being abused in that way in Nix, just because it's somewhat easier to type. For now it seems to be the pattern though, so I still think that with https://github.com/NixOS/nixpkgs/pull/215166 it's okay to replace this `if` here with an `assert assertAll isValid subpaths`

**Comment by infinisil** [comment]
This PR isn't strictly blocked by anything, so I'm going to merge this soon

**Review by roberth — APPROVED** [review]

**Pull request merged**
